### PR TITLE
Step 4: HTML Forms and POST Requests

### PR DESCRIPTION
**Pull request by leafgecko**
HTML forms to send comments as POST requests to your server

The user loads a page.

The page contains a form that posts data to the server.

The server processes or stores the data.

The server redirects back to the same page.

The page gets the processed data from the server and builds a UI.

Add elements for pokemonList in init for the showFavouritePokemon button

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    response.setContentType("text/html;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
```
Comment (fanwgwg) [inline portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: What's the reasoning behind the defaultValue? E.g, if we receive a post request for an empty comment, do we still proceed with adding this empty comment?

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -55,3 +55,12 @@ function showFavouritePokemon() {
         document.getElementById('pokemon-container').innerText = myObject;
     });
 }
+
+/* Show comments to the page */
+function showComments() {
+    fetch('/comment').then(response => response.json()) // parses the respoonse as JSON
+    .then((myObject) => { // now we can reference the fields in myObject!
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/script.js]: nit: Please have a better name for `myObject`, e.g, comments

File: portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+
+/** Servlet that processes comment. TODO: Improve formatting of response output*/
```
Comment (fanwgwg) [inline portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: Please be more specific in javadoc about the servlet.  E.g, it has post and get requests, what does each do, and what does each return.

**Review by fanwgwg — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (leafgecko) [reply portfolio/src/main/webapp/script.js]: true, fixed

File: portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java (hunk 1 above)
Comment (leafgecko) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: I would think that defaultValue is just defensive programming if something sends a POST request that is not a "comment-input". Not sure if this counts as defensive or unnecessary coding tbh.

Comment (fanwgwg) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: Yeah I agree that we should be defensive on programming. But it doesn't quite make sense (at least to me) to add an empty comment to our datastore. To be defensive, we can still have check on whether or not the comment is null/empty, but choose to no-op in such case. Or return an errro message in the response. wdyt?

Comment (leafgecko) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: Oh. so i checked the getParameter [https://docs.oracle.com/javaee/6/api/javax/servlet/ServletRequest.html#getParameter(java.lang.String)](url)
Returns the value of a request parameter as a String, or **null if the parameter does not exist**. 

I see then in this case, I do not think my getParameter method is necessary, and I can just make the getParameter within the doPost method.

Do let me know if I did that right, thank you

Comment (fanwgwg) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: Yep make sense, thanks for checking. How do you plan to handle the case if we get a null from parameter?

Comment (leafgecko) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: i'm going to do this
`    if (comment != null && !comment.isEmpty()) {
       commentList.add(comment); 
    }`

(it's going to be in my next commit that I'll push now)

Comment (leafgecko) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: > But it doesn't quite make sense (at least to me) to add an empty comment to our datastore

thank you for challenging my thinking, it really helps

Comment (fanwgwg) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: Sounds good thanks!

File: portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java (hunk 3 above)
Comment (leafgecko) [reply portfolio/src/main/java/com/google/sps/servlets/CommentProcessorServlet.java]: right added more details :)

**Review by fanwgwg — APPROVED** [review]

**Pull request merged**